// @vendure/admin-ui/esm2015/core/shared/dynamic-form-inputs/product-multi-selector-form-input/product-multi-selector-form-input.component.js
import { ChangeDetectionStrategy, ChangeDetectorRef, Component, Input } from '@angular/core';
import { DataService } from '../../../data/providers/data.service';
import { ModalService } from '../../../providers/modal/modal.service';
import { ProductMultiSelectorDialogComponent } from '../../components/product-multi-selector-dialog/product-multi-selector-dialog.component';
export class ProductMultiSelectorFormInputComponent {
    constructor(modalService, dataService, changeDetector) {
        this.modalService = modalService;
        this.dataService = dataService;
        this.changeDetector = changeDetector;
        this.mode = 'product';
        this.isListInput = true;
    }
    ngOnInit() {
        var _a, _b;
        this.mode = (_b = (_a = this.config.ui) === null || _a === void 0 ? void 0 : _a.selectionMode) !== null && _b !== void 0 ? _b : 'product';
    }
    select() {
        this.modalService
            .fromComponent(ProductMultiSelectorDialogComponent, {
            size: 'xl',
            locals: {
                mode: this.mode,
                initialSelectionIds: this.formControl.value,
            },
        })
            .subscribe(selection => {
            if (selection) {
                this.formControl.setValue(selection.map(item => this.mode === 'product' ? item.productId : item.productVariantId));
                this.changeDetector.markForCheck();
            }
        });
    }
}
ProductMultiSelectorFormInputComponent.id = 'product-multi-form-input';
ProductMultiSelectorFormInputComponent.decorators = [
    { type: Component, args: [{
                selector: 'vdr-product-multi-selector-form-input',
                template: "<div class=\"flex\">\r\n    <button (click)=\"select()\" class=\"btn btn-sm btn-secondary\">\r\n        {{ 'common.items-selected-count' | translate: { count: formControl.value?.length ?? 0 } }}...\r\n    </button>\r\n</div>\r\n",
                changeDetection: ChangeDetectionStrategy.OnPush,
                styles: [""]
            },] }
];
ProductMultiSelectorFormInputComponent.ctorParameters = () => [
    { type: ModalService },
    { type: DataService },
    { type: ChangeDetectorRef }
];
ProductMultiSelectorFormInputComponent.propDecorators = {
    config: [{ type: Input }],
    formControl: [{ type: Input }],
    readonly: [{ type: Input }]
};
