// @vendure/admin-ui/esm2015/core/shared/dynamic-form-inputs/combination-mode-form-input/combination-mode-form-input.component.js
import { ChangeDetectionStrategy, Component, Optional } from '@angular/core';
import { of } from 'rxjs';
import { map, tap } from 'rxjs/operators';
import { ConfigurableInputComponent } from '../../components/configurable-input/configurable-input.component';
/**
 * @description
 * A special input used to display the "Combination mode" AND/OR toggle.
 *
 * @docsCategory custom-input-components
 * @docsPage default-inputs
 */
export class CombinationModeFormInputComponent {
    constructor(configurableInputComponent) {
        this.configurableInputComponent = configurableInputComponent;
    }
    ngOnInit() {
        const selectable$ = this.configurableInputComponent
            ? this.configurableInputComponent.positionChange$.pipe(map(position => 0 < position))
            : of(true);
        this.selectable$ = selectable$.pipe(tap(selectable => {
            if (!selectable) {
                this.formControl.setValue(true, { emitEvent: false });
            }
        }));
    }
    setCombinationModeAnd() {
        this.formControl.setValue(true);
    }
    setCombinationModeOr() {
        this.formControl.setValue(false);
    }
}
CombinationModeFormInputComponent.id = 'combination-mode-form-input';
CombinationModeFormInputComponent.decorators = [
    { type: Component, args: [{
                selector: 'vdr-combination-mode-form-input',
                template: "<ng-container *ngIf=\"selectable$ | async; else default\">\r\n    <div class=\"btn-group btn-outline-primary btn-sm mode-select\">\r\n        <button\r\n            class=\"btn\"\r\n            (click)=\"setCombinationModeAnd()\"\r\n            [class.btn-primary]=\"formControl.value === true\"\r\n        >\r\n            {{ 'common.boolean-and' | translate }}\r\n        </button>\r\n        <button\r\n            class=\"btn\"\r\n            (click)=\"setCombinationModeOr()\"\r\n            [class.btn-primary]=\"formControl.value === false\"\r\n        >\r\n            {{ 'common.boolean-or' | translate }}\r\n        </button>\r\n    </div>\r\n</ng-container>\r\n<ng-template #default>\r\n    <small>{{ 'common.not-applicable' | translate }}</small>\r\n</ng-template>\r\n",
                changeDetection: ChangeDetectionStrategy.OnPush,
                styles: [".mode-select{text-transform:uppercase}\n"]
            },] }
];
CombinationModeFormInputComponent.ctorParameters = () => [
    { type: ConfigurableInputComponent, decorators: [{ type: Optional }] }
];
