// @vendure/admin-ui/esm2015/core/shared/dynamic-form-inputs/select-form-input/select-form-input.component.js
import { ChangeDetectionStrategy, Component, Input } from '@angular/core';
/**
 * @description
 * Uses a select input to allow the selection of a string value. Should be used with
 * `string` type fields with options.
 *
 * @docsCategory custom-input-components
 * @docsPage default-inputs
 */
export class SelectFormInputComponent {
    get options() {
        var _a;
        return ((_a = this.config.ui) === null || _a === void 0 ? void 0 : _a.options) || this.config.options;
    }
}
SelectFormInputComponent.id = 'select-form-input';
SelectFormInputComponent.decorators = [
    { type: Component, args: [{
                selector: 'vdr-select-form-input',
                template: "<select clrSelect [formControl]=\"formControl\" [vdrDisabled]=\"readonly\">\r\n    <option *ngIf=\"config.nullable\" [ngValue]=\"null\"></option>\r\n    <option *ngFor=\"let option of options\" [ngValue]=\"option.value\">\r\n        {{ (option | customFieldLabel) || option.label || option.value }}\r\n    </option>\r\n</select>\r\n",
                changeDetection: ChangeDetectionStrategy.OnPush,
                styles: ["select{width:100%}\n"]
            },] }
];
SelectFormInputComponent.propDecorators = {
    readonly: [{ type: Input }]
};
